Step3 - 구간 추가 기능 PR요청드립니다.

sangjaeoh opened the pull request:
영훈님 안녕하세요😃

지하철 구간 추가 기능 PR요청드립니다.

감사합니다 :)

sangjaeoh committed:
docs: 기능 요구사항 작성

sangjaeoh committed:
refactor: Line 엔티티 분리

Line 엔티티에서 가지고 있던 상행역, 하행역, 길이 정보를 Section 엔티티를 추가하여 이동
그에 따른 LineResponse 수정

sangjaeoh committed:
test: 인수테스트 시나리오 작성

sangjaeoh committed:
feat: 역 사이에 새로운 역을 등록하는 구간 추가 기능

sangjaeoh committed:
feat: 역 사이에 새로운 역을 등록할 경우 기존 구간 업데이트

sangjaeoh committed:
feat: 역 사이에 새로운 역을 등록할 경우 길이 체크

sangjaeoh committed:
feat: 구간 추가 시 상행역, 하행역 유효성 검사 추가

sangjaeoh committed:
refactor: 테스트코드 리팩토링

sangjaeoh committed:
fix: 구간 등록 리턴값, 에러 핸들러

sangjaeoh committed:
refactor: 도메인 리팩토링

chae-yh commented on docs/README.md:
```diff
@@ -111,4 +111,47 @@ Given 지하철 노선을 생성하고
+- 인수 조건을 검증하는 인수 테스트를 작성하세요.
+- 예외 케이스에 대한 검증도 포함하세요.
+
+### 기능 목록
```
기능목록 작성 👍 
프로그해밍 요구사항 같은 것은 따로 작성하지 않아도 될 것 같아요..!

sangjaeoh replied:
넵!

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,96 @@
+    private Station downStation;
+    private Integer distance;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
적적한 연관관계 매핑 👍

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,96 @@
+    @JoinColumn(name = "line_id")
+    private Line line;
+
+    protected Section() {
```
```suggestion
    protected Section() {}
```
[컨벤션](https://naver.github.io/hackday-conventions-java/#permit-concise-empty-block)에 따르면 이렇게 써도 될 것 같아요..!

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    private void validateDistance(Integer distance) {
+        if (distance == null || distance < 1) {
```
1은 상수 추출해주어도 좋을 것 같아요..! 그리고, validation 잘해주셨습니다 👍

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.stream.Collectors;
+
+@Embeddable
+public class Sections {
```
1급 컬렉션 👍

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+@Embeddable
+public class Sections {
+
+    @OneToMany(mappedBy = "line", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
```
적절한 연관관계 매핑 👍

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+        sections.stream().forEach(s -> s.update(section));
+    }
+
+    public List<Station> getStations() {
```
stream으로 잘 만들어주신 것 같아요 👍
그런데 이렇게 되면 역이 unique하게 나올 것 같긴한데, 순서대로 나올까요...?
관련하여 단위 테스트가 있어도 좋을 것 같습니다..!

sangjaeoh replied:
넵!

chae-yh commented on src/main/java/nextstep/subway/domain/Station.java:
```diff
@@ -1,27 +1,55 @@
         return name;
     }
+
+    @Override
```
equals와 hashCode 재정의 👍

chae-yh commented on src/main/java/nextstep/subway/dto/StationRequest.java:
```diff
@@ -10,6 +10,6 @@ public String getName() {
 
     public Station toStation() {
-        return new Station(name);
+        return Station.from(name);
```
static factory method로 잘 변경해주셨습니다 👍

chae-yh commented on src/test/java/nextstep/subway/domain/SectionTest.java:
```diff
@@ -0,0 +1,75 @@
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+public class SectionTest {
```
section은 복잡한 도메인인데 단위 테스트 잘 작성해주셨습니다 👍

chae-yh commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,172 @@
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@DisplayName("구간 관련 기능")
+public class SectionAcceptanceTest extends BaseAcceptanceTest {
```
꼼꼼한 인수테스트 👍

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+
+    public void add(Section section) {
+        validateAddContainsAllStation(section);
+        validateAddNonContainsStation(section);
```
최초에 노선이 만들어지고, section이 하나도 없을 때 A-B 구간을 등록하려고 하면 여기서 에러가 나지 않을까요..?

sangjaeoh replied:
최초에 노선을 만들 때, 구간이 무조건 받게 되어 있어서 기존에 오류가 나진 않았지만
방어코드를 추가하도록 하겠습니다!

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    public void update(Section section) {
```
물론 그런 값이 들어오지는 않겠지만, 형태상으로 봤을 때 동일한 Section인데 길이가 더 짧다면 업데이트가 두 번 일어날 것 같아요. 이런 경우를 대비해서 validation체크를 해줄 수 있지 않을까요..?

sangjaeoh replied:
제가 생각하기에는
```java
// Sections
public void add(Section section) {
        validateAddContainsAllStation(section);
        validateAddNonContainsStation(section);
        ...
```
여기서 유효성 검사를 하기 때문에 동일한 섹션(상행역, 하행역 일치)이 들어올 수 없을것 같습니다 🤔

그리고 업데이트는 
```java
// Section
public void update(Section section) {
        if (this.upStation.equals(section.upStation)) {
        }
        if (this.downStation.equals(section.downStation)) {
        }
    }
``` 
상행역은 상행역, 하행역은 하행역이 같을 때만 동작하기 때문에
모든 구간 중 0번이거나 1번만 업데이트 하게 됩니다!

chae-yh replied:
넵 저도 말씀하신 내용 동의합니다..! 다만 걱정이 되었던 부분은 1) Seciton의 update가 public으로 열려있고 2) 말씀하신 validation은 Sections에서 Section을 update 해줄 때만 검증이되고 직접 Section의 update를 실행할 경우에는 방어가 안될 것이라 생각했습니다...!

chae-yh commented on src/test/java/nextstep/subway/domain/LineTest.java:
```diff
@@ -61,6 +61,28 @@ public class LineTest {
 
+    @Test
+    @DisplayName("잠실역, 장지역 구간 사이에 문정역 구간을 추가한다")
+    void addSection() {
```
정확히 같은 형태는 아니지만, 인수테스트 시에 
given -> when -> then -> given -> when -> then을 반복하며 테스트해야 하는 경우가 있는데, 이런 경우는 [DynamicTest](https://tecoble.techcourse.co.kr/post/2020-07-31-dynamic-test/)를 써봐도 좋을 것 같습니다 ㅎㅎ

sangjaeoh replied:
이런것이 있는줄 몰랐습니다. 😂
링크 주신 블로그의 설명처럼 `@BeforeEach`나 `@AfterEach` 사용하지 않는 테스트 중에
`given -> when -> then -> given -> when -> then`이 반복된다면 사용해 보도록 하겠습니다!
감사합니다 👍

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private void updateSections(Section section) {
```
이 방법도 재밌네요 👍

chae-yh reviewed (changes requested):
상재님 안녕하세요. step3 코드도 잘 작성해주셨습니다 👍 다만 조금 더 고민해주시면 좋을만한 내용이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

sangjaeoh committed:
style: 코드 컨벤션 적용

sangjaeoh committed:
refactor: PR 수정

- distance 유효성 검사의 최소 길이를 상수로 변경
- Sections add() 시 유효성 검사에 방어코드 추가
- LineResponse의 지하철 역 값들을 정렬 및 테스트 코드 수정

chae-yh commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    private void validateStations(Station... stations) {
+        if (stations == null || Arrays.stream(stations).anyMatch(Objects::isNull)) {
```
```suggestion
        if (Objects.isNull(stations) || Arrays.stream(stations).anyMatch(Objects::isNull)) {
```
이 부분은 중요한 것은 아니지만 이렇게 써줄 수도 있을 것 같아요..!

chae-yh commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private void updateSections(Section section) {
+        sections.stream().forEach(s -> s.update(section));
```
s와 같은 단축어보다는 해당 변수를 잘 표현하는 단어를 써도 좋을 것 같습니다. ㅎㅎ

chae-yh commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -1,32 +1,32 @@
+        this.stations = sections.getStations()
+                .stream()
+                .map(StationResponse::from)
+                .sorted(Comparator.comparing(StationResponse::getId))
```
음.. station의 id 순이면, 해당 노선에서 상행 종점부터 하행 종점까지 순서대로 들어갈까요..? 예를 들어, 특정 역이 환승역일 때면 어떻게 될까요..?
그리고 역을 순서대로 구하는 것은 dto의 역할이라기 보다는, 도메인의 역할은 아닐까요...?

chae-yh reviewed (approved):
상재님 피드백 잘 반영해 주셨습니다 👍 조금 더 고민해주시면 좋을만한 내용이 있어 코멘트 드렸지만 전체적인 미션의 목적은 달성하였다고 생각이 들어 merge하겠습니다.

벌써 마지막 step이네요 파이팅 입니다 : )
